### ISSUE TITLE
Kotlin compiler daemon keeps disappearing with file-system watching enabled on macOS

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#13382]
Copied over from the forum by @MyDogTom: https://discuss.gradle.org/t/build-slowdown-after-enabling-file-system-watching/36415/3

----

I enabled file-system watching by adding `org.gradle.unsafe.watch-fs=true` in global `gradle.properties`.

After that I see the following messages in console:

`Could not connect to kotlin daemon. Using fallback strategy.` - this happens for every module.

```
RMI TCP Accept-0: accept loop for ServerSocket[addr=localhost/127.0.0.1,localport=50888] throws
java.net.SocketException: Socket closed
        at java.net.PlainSocketImpl.socketAccept(Native Method)
        at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
        at java.net.ServerSocket.implAccept(ServerSocket.java:560)
        at java.net.ServerSocket.accept(ServerSocket.java:528)
        at sun.rmi.transport.tcp.TCPTransport$AcceptLoop.executeAcceptLoop(TCPTransport.java:405)
        at sun.rmi.transport.tcp.TCPTransport$AcceptLoop.run(TCPTransport.java:377)
        at java.lang.Thread.run(Thread.java:748)
```

Build succeeded, but it took much longer than usual.

MacOS 10.15.4
Java 1.8.0_232
Gradle 6.5
Kotlin 1.3.72

Only one daemon is running and it was started after changing `gradle.properties`.

----

I found quite a few Kotlin compiler logs. There seems to be a repetitive pattern. First daemon starts and shutdown after a few seconds. Meanwhile another another daemon starts, detects previous daemon and shutdown.
```
2020-06-09 14:29:53.862 [daemon] INFO: Kotlin compiler daemon version 1.3.61-release-180 (1.3.61)
2020-06-09 14:29:56.607 [daemon] INFO: daemon JVM args: -Djava.awt.headless=true -Djava.rmi.server.hostname=127.0.0.1 -Xmx8g -XX:MaxMetaspaceSize=1024m -Dkotlin.environment.keepalive -ea
2020-06-09 14:29:56.624 [daemon] INFO: daemon args: --daemon-runFilesPath /Users/svyatoslav.chatchenk/Library/Application Support/kotlin/daemon --daemon-autoshutdownIdleSeconds=7200 --daemon-compilerClasspath /Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-compiler-embeddable/1.3.61/c35c0833d5a1e3d229359758482c480306724618/kotlin-compiler-embeddable-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-reflect/1.3.61/2e07c9a84c9e118efb70eede7e579fd663932122/kotlin-reflect-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.3.72/8032138f12c0180bc4e51fe139d4c52b46db6109/kotlin-stdlib-1.3.72.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-script-runtime/1.3.61/e5feb39c8de63dcbdc1154b34ca4413874231605/kotlin-script-runtime-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-daemon-embeddable/1.3.61/4de83e41a92c8395d89ab2e81d26b788ba163824/kotlin-daemon-embeddable-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.intellij.deps/trove4j/1.0.20181211/216c2e14b070f334479d800987affe4054cd563f/trove4j-1.0.20181211.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.3.72/6ca8bee3d88957eaaaef077c41c908c9940492d8/kotlin-stdlib-common-1.3.72.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/16.0.2/dab5655eb9abe23f2245c34e66c9813c7daf4563/annotations-16.0.2.jar:/Library/Java/JavaVirtualMachines/amazon-corretto-8.jdk/Contents/Home/lib/tools.jar
2020-06-09 14:29:56.683 [daemon] INFO: starting daemon
2020-06-09 14:29:57.023 [daemon] INFO: daemon is listening on port: 17284
2020-06-09 14:29:57.032 [compiler] INFO: initiate elections
2020-06-09 14:29:58.022 [compiler] INFO: No more clients left
2020-06-09 14:29:58.023 [compiler] INFO: Delayed shutdown in 1000ms
2020-06-09 14:29:58.078 [compiler] INFO: getDaemonJVMOptions: DaemonJVMOptions(maxMemory=8g, maxMetaspaceSize=1024m, reservedCodeCacheSize=, jvmParams=[-Djava.awt.headless=true, -Djava.rmi.server.hostname=127.0.0.1, -Dkotlin.environment.keepalive, -ea, Dkotlin.environment.keepalive, ea])
2020-06-09 14:29:59.024 [compiler] INFO: Shutdown started
2020-06-09 14:29:59.025 [compiler] INFO: Memory stats: total: 245mb, free: 234mb, max: 7282mb
2020-06-09 14:29:59.026 [compiler] INFO: Shutdown complete
```
```
2020-06-09 14:29:57.390 [daemon] INFO: Kotlin compiler daemon version 1.3.61-release-180 (1.3.61)
2020-06-09 14:29:57.604 [daemon] INFO: daemon JVM args: -Djava.awt.headless=true -Djava.rmi.server.hostname=127.0.0.1 -Xmx8g -XX:MaxMetaspaceSize=1024m -Dkotlin.environment.keepalive -ea
2020-06-09 14:29:57.626 [daemon] INFO: daemon args: --daemon-runFilesPath /Users/svyatoslav.chatchenk/Library/Application Support/kotlin/daemon --daemon-autoshutdownIdleSeconds=7200 --daemon-compilerClasspath /Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-compiler-embeddable/1.3.61/c35c0833d5a1e3d229359758482c480306724618/kotlin-compiler-embeddable-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-reflect/1.3.61/2e07c9a84c9e118efb70eede7e579fd663932122/kotlin-reflect-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.3.72/8032138f12c0180bc4e51fe139d4c52b46db6109/kotlin-stdlib-1.3.72.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-script-runtime/1.3.61/e5feb39c8de63dcbdc1154b34ca4413874231605/kotlin-script-runtime-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-daemon-embeddable/1.3.61/4de83e41a92c8395d89ab2e81d26b788ba163824/kotlin-daemon-embeddable-1.3.61.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.intellij.deps/trove4j/1.0.20181211/216c2e14b070f334479d800987affe4054cd563f/trove4j-1.0.20181211.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.3.72/6ca8bee3d88957eaaaef077c41c908c9940492d8/kotlin-stdlib-common-1.3.72.jar:/Users/svyatoslav.chatchenk/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/16.0.2/dab5655eb9abe23f2245c34e66c9813c7daf4563/annotations-16.0.2.jar:/Library/Java/JavaVirtualMachines/amazon-corretto-8.jdk/Contents/Home/lib/tools.jar
2020-06-09 14:29:57.733 [daemon] INFO: starting daemon
2020-06-09 14:29:57.984 [daemon] INFO: daemon is listening on port: 17198
2020-06-09 14:29:57.996 [compiler] INFO: initiate elections
2020-06-09 14:29:58.009 [compiler] INFO: found daemon on port 17284 (1000 ms old), trying to connect
2020-06-09 14:29:58.096 [compiler] INFO: Assuming other daemons have higher prio, handover clients to it and schedule shutdown: my runfile: kotlin-daemon.2020-06-09T12-29-57.870Z.40c8bc1be5268d9e26a826a2aae8bb32.17198.run (1591705797000) vs best other runfile: kotlin-daemon.2020-06-09T12-29-56.921Z.40c8bc1be5268d9e26a826a2aae8bb32.17284.run (1591705796000)
2020-06-09 14:29:58.097 [compiler] INFO: Graceful shutdown signalled
2020-06-09 14:29:58.099 [compiler] INFO: Delayed shutdown in 1000ms
2020-06-09 14:29:59.100 [compiler] INFO: Shutdown started
2020-06-09 14:29:59.102 [compiler] INFO: Memory stats: total: 245mb, free: 235mb, max: 7282mb
2020-06-09 14:29:59.102 [compiler] INFO: Shutdown complete
```

This process. There are at least a hundred of logs like that. Btw full message is
```
Could not perform incremental compilation: Could not connect to Kotlin compile daemon
Could not connect to kotlin daemon. Using fallback strategy.
```

>Can you tell us some more about your build and your environment?

This is an android application with more than 100 Gradle modules. Let me know what exactly you want to know. I could share a build scan report with you privately. Let me know if it could help and how to send it to you.

>Can you maybe even provide a reproducer?

I'll try to create one later.

---
cc: @gradle/execution

**Comment by lptr** [comment]
This is a build scan from the user from the forum: https://scans.gradle.com/s/bqoocxionmtcc/console-log#L388

**Comment by lptr** [comment]
More evidence pointing at sockets conflicting with file-system watching somehow; a user reports:

> I update our project to gradle 6.5 and started using file watching locally. After I change the android gradle plugin, I got dependency download errors. Socket closed. I killed daemon, rebuild but buiid failed again with Socket closed. I disabled file watch, killed daemon and build succeed. Now I have succeed build, I am not sure if I can reproduce, but I probably can. How can I collect more information to create a issue or is here right place to report?

**Comment by kozaxinan** [comment]
I am the user who reported to @lptr . Today, I updated a dependency (ktlint) and build fails with socket exception while downloading new dependency.

Here are the full log of `./gradle ktlintCheck --stacktrace`
https://gist.github.com/kozaxinan/5fafea7f97bfd15f924068aaa15a43cd

**Comment by cesar1000** [comment]
I’ve spent some more time looking into this issue and, based on the timestamps in the logs and the lack of activity on Wireshark, I feel that the daemon takes took longer than a second to send a first request to the Kotlin daemon, by which time the daemon has initiated shutdown. It doesn't look like this wait time can be configured programmatically. My hypothesis is that the daemon rejects requests coming after shutdown has started, but I haven't been able to verify.

One consistent observation is that requests seem to be delayed significantly whenever using filesystem watching, so that looks like it may be related. I can repro consistently, so I'm happy to collect more info if you have any suggestions.

**Comment by lptr** [comment]
Thanks for investigating @cesar1000. Our current hypothesis is that watching many (100+) hierarchies on macOS slows all I/O down somehow, and that's causing certain sensitive things to timeout, like the Kotlin daemon.

FTR, we are currently watching many locations instead of the necessary few because of a bug we are about to fix soon (see #13222) on `master`. We expect the fix for that bug will also fix the problem with the Kotlin compiler daemon.

**Comment by lptr** [comment]
I've been trying to validate this idea that watching lots of directories leads to sockets being slowed down. I created a small test that opened a socket connection to another process running on the same host, sent and received a bunch of text messages, then tore down the connection. I tried running this while watching 0, 1, 100 and 500 directories, but always got the following characteristics:

- first connection takes 20-40 ms
- re-connecting takes 1 ms
- sending and receiving messages takes ~0.22 ms on average, with one or two outliers around 5-8 ms

These parameters don't seem to change based on how many directories are being watched. :(

**Comment by lptr** [comment]
I've retried with Java RMI, the protocol Kotlin uses to connect to compiler daemons. I set up a very simple protocol based on this example, and kept a client process running that would open a new RMI connection to the server process, receive a message, wait 10 ms, then start from the beginning ad infinitum.

I tested this setup with the following kinds of load, measuring the roundtrip time.

0) No additional load.
1) Heavy CPU-only load in another process putting random items into a `ConcurrentHashMap` on 100 threads concurrently.
2) Mostly I/O load in another process running 100 threads creating files, writing to them, and then deleting them.
3) The same as 2), but also watching the directories being modified, stopping the watching in the middle of the work.

The roundtrip time measured in 0) was mostly 1 or 2 ms, with occasional jumps as high as 20 ms a couple times each minute. With 1) and 2) running in parallel there wasn't any significant difference in the roundtrip times.

However, with watching enabled in 3), the roundtrip time kept jumping up to as high as 150 ms every few seconds. I tried with 2 threads modifying 50 watched directories, 10 threads vs 10 watched directories, and finally 50 threads modifying 2 watched directories. The effect was most pronounced in the 10 vs 10 case, but noticeable in the other two extremes, too.

One other thing I noticed that my bluetooth mouse started to become choppy with 3) running, something I haven't noticed with the other scenarios.

I think from this we can conclude that there is some kind of relationship between socket connection speeds and watching many heavily modified directories as the same time. I'm guessing a more complicated RMI protocol (like the one used by Kotlin daemons) would be more affected, and thus it is entirely possible that we see a roundtrip that takes longer than a second, which is the timeout value used by the Kotlin daemon.
